Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Github5623.xaml.cs

```diff
@@ -0,0 +1,168 @@
+﻿using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Threading;
+using System.Threading.Tasks;
+using Xamarin.Forms.CustomAttributes;
+using Xamarin.Forms.Internals;
+using Xamarin.Forms.Xaml;
+
+#if UITEST
+using Xamarin.UITest;
+using NUnit.Framework;
+using Xamarin.Forms.Core.UITests;
+using System.Linq;
+#endif
+
+namespace Xamarin.Forms.Controls.Issues
+{
+#if UITEST
+	[Category(UITestCategories.CollectionView)]
+#endif
+#if APP
+	[XamlCompilation(XamlCompilationOptions.Compile)]
+#endif
+	[Preserve(AllMembers = true)]
+	[Issue(IssueTracker.Github, 5623, "CollectionView with Incremental Collection (RemainingItemsThreshold)", PlatformAffected.All)]
+	public partial class Github5623 : TestContentPage
+	{
+		int _itemCount = 100;
+		int _maximumItemCount = 1000;
+		int _pageSize = 100;
+		static SemaphoreSlim SemaphoreSlim = new SemaphoreSlim(1, 1);
+
+		public Github5623()
+		{
+#if APP
+			Device.SetFlags(new List<string> { CollectionView.CollectionViewExperimental });
+
+			InitializeComponent();
+
+			BindingContext = new ViewModel5623();
+#endif
+		}
+
+		protected override void Init()
+		{
+
+		}
+
+		private async void CollectionView_RemainingItemsThresholdReached(object sender, System.EventArgs e)
+		{
+			await SemaphoreSlim.WaitAsync();
+			try
+			{
+				var itemsSource = (sender as CollectionView).ItemsSource as ObservableCollection<Model5623>;
+
+				if (itemsSource.Count == _maximumItemCount)
+				{
+					System.Diagnostics.Debug.WriteLine("Count: " + itemsSource.Count);
+					return;
+				}
+
+				var nextSet = await GetNextSetAsync();
+
+				Device.BeginInvokeOnMainThread(() =>
+				{
+					foreach (var item in nextSet)
+					{
+						itemsSource.Add(item);
```

Comment:
While the web service is async, items must be added to the underlying data source on the main thread. Sometimes there is a small lag when that happens especially when scrolling is continuing. Is there any way to prevent this?

Reply:
@adrianknight89 Did this ever get resolved? Been trying to optimize CarouselView content loading for a while now, especially when loading/unloading content on the fly, but I still have a split second of blocking UI thread lag.